Name: __________________        Class: __________________     Date: __________________

<center>
    <img width="100%" src="https://raw.githubusercontent.com/astroDimitrios/Astronomy/main/APBanner.png" alt='A&P Logo'>
</center>

# Mars 2020 - Hohmann Orbits

### AIM - Determine the next launch window to Mars
Difficulty: Easy

## Start:

In this activity you are going to look at how scientists get rovers like Perseverance to Mars. This notebook is adapted from the NASA activity found [here](https://www.jpl.nasa.gov/edu/teach/activity/lets-go-to-mars-calculating-launch-windows/?utm_source=iContact&utm_medium=email&utm_campaign=nasajpl-edu&utm_content=m2msc_week2).

<center>
    <img width="50%" src="./Images/perseverance_artist.png" alt='Perseverance rover'>
</center>

## Contents

* [Start](#Start)
* [Assumptions](#ASS)
* [Heliocentric Longitude](#LONG)
* [Semi-major axis of the Orbit](#CALCA)
* [Time to Mars](#TIME)
* [The next Launch Window](#LAUNCH)
* [Where to next?](#WHERE)
* [Discussion](#DISS)
* [Extension](#OTY)
* [References](#REF)


## Let's Go: <a class="anchor" id="Start"></a>

To get a rover to Mars you first have to fire your rocket so that it escapes the Earth's gravitational pull. The rover then coasts in its housing to Mars and fires thrusters to slow down and land on the Martian surface at the end of its journey.

The most energy efficient way of getting our rover to Mars is with a Hohmann transfer.
A Hohmann transfer between the Earth and Mars is shown in the image below.

<center>
    <img width="75%" src="./Images/hohmann_transfer.jpg" alt='hohmann transfer'>
</center>

In this image the blue circle is the Earth's orbit, the red circle is Mars' orbit, and the grey line is the path taken by the spacecraft to Mars.

Notice that there is also a dashed grey line in this image. The two grey lines together form the Hohmann orbit. So a Hohmann transfer is one half of a Hohmann orbit!

For our transfer to work the spacecraft must leave Earth at the Hohmann orbit's perihelion (the point on the orbit when you're closest to the Sun) and arrive at Mars at the Hohmann orbit's aphelion (the point on the orbit when you're furthest from the Sun).

<div class="alert alert-block alert-info">
    <p><strong>Assumptions</strong> </p><a class="anchor" id="ASS"></a>

Before you go ahead and do any calculations we have to outline some assumptions:

* The orbits of Earth and Mars are circular (they are in reality elliptical)
* Earth and Mars travel at constant speeds (they don't - remember Kepler's 3rd law!)
* The orbits of Earth and Mars are in the same plane (there is a small difference in their inclinations)</div>

**Heliocentric Longitude** <a class="anchor" id="LONG"></a>

We will be using heliocentric longitudes when talking about the positions of Earth and Mars. A heliocentric longitude is the position of an object with repsect to the Sun (hence helio - Sun, centric - centre). It is measured from the primary direction, which is the position of the Earth at the vernal equinox, around the ecliptic plane counterclockwise. The ecliptic plane is an imaginary disk swept out by a line connecting the Earth and the Sun as the Earth completes one orbit.

This means at launch the Earth is at a longitude of $0^{\circ}$ and at landing Mars is at $180^{\circ}$. This is shown in the image below which looks down on the ecliptic plane.

<center>
    <img width="75%" src="./Images/helio_long_2.jpg" alt='heliocentric longitudes'>
</center>

**Calculating the Semi-Major Axis** <a class="anchor" id="CALCA"></a>

Time to code! The first thing you will do is calculate the semi major axis $a$ of the transfer orbit. To do this you can use the equation below:

$$a = \frac{1}{2}(r_{Earth} + r_{Mars})$$

where $r_{Earth}$ is the average distance from the Sun to the Earth or $1\ AU$, and $r_{Mars}$ is the average distance from the Sun to Mars which is $1.52\ AU$. Remember $AU$ stands for Astronomical Unit and $1\ AU =  1.495978707\times 10^{11}\ m$.

<div class="alert alert-block alert-success">
In the code cell below:

* Enter the values for $r_{Earth}$ and $r_{Mars}$
* Calculate $a$
* Print the value of $a$ and its unit ($Au$)</div>

In [1]:
rEarth = 1                   # AU
rMars = 1.52                 # AU
a = .5*(rEarth+rMars)        # AU
print('a = {} AU'.format(a))

a = 1.26 AU


**Time to Mars** <a class="anchor" id="TIME"></a>

To calculate the travel time to Mars we need to calculate the orbital period of the Hohmann orbit (the time for one orbit). You can do this with Kepler's 3rd law which states that the orbital period, $P$ is proportional to the semi-major axis $a$. The law is represented by the equation:

$$P^2 = ka^3$$

Here $k$ is a constant you will work out for Earth first and then you can use that value in your Hohmann orbit calculation like this:

$$k = \frac{P_{Earth}^2}{a_{Earth}^3}$$

$$P = \sqrt{ka^3}$$

To simpify matters use $P_{Earth} = 1\ year$ and $a_{Earth} = 1\ AU$.
<div class="alert alert-block alert-success">
In the code cell below:

* Enter the value for k
* Calculate P
* Print the value of k and its unit
* Print the value of P and its unit</div>

In [2]:
import numpy as np
from IPython.display import display, Math

k = 1                # years^2 / AU^3
P = np.sqrt(k*a**3)   # years

display(Math(r'k = {}\ years^2 / a^3'.format(k)))
display(Math(r'P = {}\ years'.format(P)))

<IPython.core.display.Math object>

<IPython.core.display.Math object>

<div class="alert alert-block alert-success">In the next code cell lets convert $P$ into days. Then all we have to do is half the period $P$ to get the travel time to Mars. Remember this is because our Hohmann transfer is half a full Hohmann orbit.

* Convert $P$ to days (use $1\ year = 365.25\ days$)
* Half the period to get the travel time <strong>t_toMars</strong>
* Print out a rounded value for <strong>t_toMars</strong> and its unit</div>

In [3]:
P_days = P*365.25      # days
t_toMars = P_days/2    # days
display(Math(r'Time\ to\ Mars = {:.1f}\ days'.format(t_toMars)))

<IPython.core.display.Math object>

Amazing! You have just calculated the time it takes for our spaceraft to travel to Mars carrying the rover. Now we can work out the position Mars must be in at launch so that it has a longiutude of $180^{\circ}$ when the spaceraft arrives.

To do this we just need to work out how far Mars moves in degrees over our $\approx 259$ day flight time and subtract that from the arrival longitude ($180^{\circ}$).
<div class="alert alert-block alert-success">
In the code cell below:

* Calculate how many degrees Mars moves a day assuming it orbits the Sun in $687$ days
* Times this value by the <strong>t_toMars</strong> variable to get how many degrees Mars has moved since launch - store this in the <strong>mars_moved</strong> variable
* Subtract <strong>mars_moved</strong> from $180^{\circ}$ to get our launch longitude
* Print out a rounded value for the launch longitude and its unit</div>

In [4]:
mars_moved_day = 360/687                  # degrees / day
mars_moved = mars_moved_day * t_toMars    # degrees
launch_long = 180 - mars_moved            # degrees
display(Math(r'Mars\ Launch\ Longitude: {:.1f}^{{\circ}}'.format(launch_long)))

<IPython.core.display.Math object>

This means that Mars must be $44.6^{\circ}$ ahead of the Earth at launch for a successful Hohmann transfer as shown in the image below:

<center>
    <img width="75%" src="./Images/mars_launch.jpg" alt='mars launch angle'>
</center>

The last thing to do is work out when Mars will be $44.6^{\circ}$ ahead of the Earth in the future.

**Calculating the Next Launch Window** <a class="anchor" id="LAUNCH"></a>

I have included a file called **earth_mars_long.csv** with this notebook. You should see it on the filetree to the left of the screen. It contains heliocentric longitudes for Earth and Mars from 2010 up to 2030. The data was generated using [this](https://omniweb.gsfc.nasa.gov/coho/helios/heli.html) NASA website.

Run the next code cell to load the data and see what's inside using the Pandas library.

In [5]:
# run me
import pandas as pd

df = pd.read_csv('earth_mars_longs.csv')
df.head()

YEAR  DAY  EARTH_RAD_AU  EARTH_SE_LAT  EARTH_SE_LON  MARS_RAD_AU  \
0  2010    1          0.98             0         100.5         1.63   
1  2010    2          0.98             0         101.5         1.63   
2  2010    3          0.98             0         102.5         1.63   
3  2010    4          0.98             0         103.5         1.63   
4  2010    5          0.98             0         104.5         1.63   

   MARS_SE_LAT  MARS_SE_LON  
0          1.7        116.8  
1          1.7        117.2  
2          1.7        117.7  
3          1.7        118.2  
4          1.7        118.6

You can see we have 8 columns. The year is followed by the day of the year. Then we have the orbital radius, heliocentric latitude, and heliocentric longitude for the Earth. Then the same for Mars.

Lets add another column which is the longitude for Mars minus the longitude for the Earth. Then we can select the dates where this column is $\approx 44.6^{\circ}$.

Remember data in a pandas ```df``` is accessed like this:

```Python
years = df['YEAR']
```

and new columns made like this:

```Python
df['NEW_COL'] = df['YEAR'] + df['DAY']
```
In this example the new column would be the sum of the year and the day!
<div class="alert alert-block alert-success">
In the code cell below:

* Add a new column called LON_DIFF which is MARS_SE_LON minus EARTH_SE_LON
* Use <strong>df.head()</strong> like we did above to see the new column</div>

In [6]:
df['LON_DIFF'] = df['MARS_SE_LON'] - df['EARTH_SE_LON']
df.head()

YEAR  DAY  EARTH_RAD_AU  EARTH_SE_LAT  EARTH_SE_LON  MARS_RAD_AU  \
0  2010    1          0.98             0         100.5         1.63   
1  2010    2          0.98             0         101.5         1.63   
2  2010    3          0.98             0         102.5         1.63   
3  2010    4          0.98             0         103.5         1.63   
4  2010    5          0.98             0         104.5         1.63   

   MARS_SE_LAT  MARS_SE_LON  LON_DIFF  
0          1.7        116.8      16.3  
1          1.7        117.2      15.7  
2          1.7        117.7      15.2  
3          1.7        118.2      14.7  
4          1.7        118.6      14.1

Scroll to the right if you can't see the column. On January the 1st 2010 Mars was $\approx 16.3^{\circ}$ ahead of the Earth!

All that's left for you to do is find the dates where this new ```LON_DIFF``` column is $\approx 44.6^{\circ}$.

To find rows based on a column value we can use the ```loc``` function like this:

```Python
df_years = df.loc[(df['MARS_RAD_AU'] >= 1.50) & (df['MARS_RAD_AU'] <= 1.52)]
df_years
```

This snippet would create a new dataframe called ```df_years``` which includes all the rows of our orginal dataframe where the column ```MARS_RAD_AU``` had values between $1.50$ and $1.52\ AU$.

<div class="alert alert-block alert-success">
In the code cell below:

* Create a new dataframe ```df_launch``` which has all the rows of the original dataframe where the ```LON_DIFF``` column has values between $44.0^{\circ}$ and $45.0^{\circ}$ (not exactly $44.6^{\circ}$ due to errors from our assumptions!)
* Print this new dataframe
* When is the next launch window according to your new table?
    </div>

In [7]:
df_launch = df.loc[(df['LON_DIFF'] >= 44.0) & (df['LON_DIFF'] <= 45.0)]
df_launch

YEAR  DAY  EARTH_RAD_AU  EARTH_SE_LAT  EARTH_SE_LON  MARS_RAD_AU  \
715   2011  351          0.98             0          84.7         1.65   
716   2011  352          0.98             0          85.7         1.65   
1479  2014   19          0.98             0         118.8         1.66   
1480  2014   20          0.98             0         119.8         1.66   
2243  2016   53          0.99             0         152.8         1.62   
2244  2016   54          0.99             0         153.8         1.62   
3015  2018   94          1.00             0         194.2         1.53   
3016  2018   95          1.00             0         195.2         1.53   
3806  2020  155          1.01             0         252.9         1.41   
3807  2020  156          1.01             0         253.8         1.41   
3808  2020  157          1.01             0         254.8         1.41   
5409  2024  297          0.99             0          30.1         1.53   
5410  2024  298          0.99             0          31.1         1.53   
6180  2026  337          0.99             0          70.8         1.62   
6181  2026  338          0.99             0          71.8         1.63   
6945  2029    6          0.98             0         106.0         1.66   
6946  2029    7          0.98             0         107.0         1.67   

      MARS_SE_LAT  MARS_SE_LON  LON_DIFF  
715           1.8        129.5      44.8  
716           1.8        129.9      44.2  
1479          1.7        163.3      44.5  
1480          1.7        163.8      44.0  
2243          1.0        197.7      44.9  
2244          1.0        198.1      44.3  
3015         -0.3        239.0      44.8  
3016         -0.3        239.5      44.3  
3806         -1.7        297.9      45.0  
3807         -1.7        298.5      44.7  
3808         -1.7        299.1      44.3  
5409          0.8         75.0      44.9  
5410          0.8         75.5      44.4  
6180          1.7        115.7      44.9  
6181          1.7        116.2      44.4  
6945          1.8        150.5      44.5  
6946          1.8        151.0      44.0

Do you see a pattern? Launch windows occur every two years and two months (specifically 780 days). You can see this cycle in the image below from NASA/JPL:

<center>
    <img width="75%" src="./Images/NASAJPL_windows.jpg" alt='launch windows'>
</center>

In this image you can see the distance between Earth and Mars plotted as a blue line. The launch windows occur just before Mars and the Earth are closest to each other. The dark blue triangle in 2020 is the Mars 2020 mission carrying the Perseverence rover! The Rosalind Franklin/ExoMars rover has been pushed to launch in 2022 during the next window.

**Where to Next** <a class="anchor" id="WHERE"></a>

Have a think about some of the questions in the discussion section below.

Do some of the extension tasks below.

Watch these awesome videos:

* [Perseverance Launch to Mars](https://www.youtube.com/watch?v=JIB3JbIIbPU) on 30th July 2020 by NASA
* [Getting to Mars](https://www.youtube.com/watch?list=PL9TFrgFq7557nWqmfuVngU22OhTpUE9gg&v=cA8ar8vwpTs&feature=emb_logo) by NASA/JPL
* [Mars Launch Windows](https://www.youtube.com/watch?v=p4re3LCkNcg) by CGTN 
* [Perseverance Rover Landing Animation](https://www.youtube.com/watch?v=rzmd7RouGrM) by NASA/JPL

## Saving your Work

Always save your work at the end and download the **.ipynb** and other files you need to keep!

## Discussion <a class="anchor" id="DISS"></a>

* Do we have to launch the spacecraft on the exact day we found above? What happens if we launch early or late?
* Research: What is the average length of a launch window to Mars.

## Extensions <a class="anchor" id="OTY"></a>

Have a go at the tasks below. I haven't provided solutions and there are probably a few ways to solve each one! Remember if you are stuck you can consult the Python and NumPy docs at https://docs.python.org/3/, and https://numpy.org/doc/stable/, find help from one of these websites www.w3schools.com, www.python.org, www.learnpython.org or you can also get advice from www.stackoverflow.com. You can also ask your instructor for help or email me at astrodimitrios@gmail.com.

**Task 1:** When was the launch window before 2020? What missions were launched during this window, are they still working, and what did they discover?    
**Task 2:** Alter the values of $44.0^{\circ}$ and $45.0^{\circ}$ which we used to select our launch windows. How much wider does this range have to be to see the 2022 launch window?    
**Task 3:** Add some code below to work out where the Earth is in its orbit when the rover arrives at Mars.  
**Task 4:** See if you can adapt your code to calculate a Hohmann transfer to Saturn or any other planet.    
**Task 5:** Create an animation of a spaceship travelling on a Hohmann transfer to Mars using Matplotlib! 

In [8]:
# Your code here:


## References <a class="anchor" id="REF"></a>

Shamelessley adapted from the NASA/JPL activity Let's Go to Mars! Calculating Launch Windows found [here](https://www.jpl.nasa.gov/edu/teach/activity/lets-go-to-mars-calculating-launch-windows/?utm_source=iContact&utm_medium=email&utm_campaign=nasajpl-edu&utm_content=m2msc_week2).

All images are from NASA/JPL.

Data taken from the HelioWeb NASA page [here](https://omniweb.gsfc.nasa.gov/coho/helios/heli.html). 

## Sharing

If you share, use or modify this activity in any way use the citation in this [txt file](https://github.com/astroDimitrios/Astronomy/blob/master/CITATION.txt).    
Please contact me at astrodimitrios@gmail.com with any suggestions, mistakes found, or general questions about teaching astronomy with Python.

© Dimitrios Theodorakis GNU General Public License v3.0 

https://github.com/astroDimitrios/Astronomy  